archive/imo/imo2021_q1.lean

```diff
@@ -0,0 +1,264 @@
+  suffices : ∃ (l : ℤ), l ≥ 0 ∧ 2 * (l : ℝ) ≥ 2 + sqrt (4 + 2 * n) ∧ (l : ℝ) ≤ sqrt (1 + n) - 1,
+  cases this with l t,
+  use l,
+  refine ⟨t.1, ⟨lower_bound n l t.2.1, upper_bound n l t.1 t.2.2⟩⟩,
+  use floor (sqrt (1 + n) - 1),
+  split,
+  { simp only [ge_iff_le],
+    rw floor_nonneg,
+    norm_num,
+    rw le_sqrt,
+    repeat {norm_cast, linarith}},
+  refine ⟨_, floor_le (sqrt (1 + n) - 1)⟩,
+  { suffices : 2 * (sqrt (1 + n) - 2) ≥ 2 + sqrt (4 + 2 * n),
+    apply le_trans this _,
+    simp only [mul_le_mul_left, zero_lt_bit0, zero_lt_one],
+    suffices : sqrt (1 + n) - 1 ≤ ↑⌊sqrt (1 + n) - 1⌋ + 1,
+    linarith,
+    have t :  (⌈sqrt (1 + n) - 1⌉:ℝ) ≤ ⌊sqrt (1 + n) - 1⌋ + 1,
+    norm_cast,
+    exact ceil_le_floor_add_one _,
+    apply le_trans _ t,
+    exact le_ceil (sqrt (1 + ↑n) - 1),
+    suffices : sqrt (4 + 2 * n) ≤ 2 * (sqrt (1 + n) - 3),
+    linarith,
+    exact radical_inequality hn },
+end
+
+/-
+We show there exists a triplet a, b, c ∈ [n , 2n] with a < b < c and each of the sums (a + b),
+(b + c), (a + c) being a perfect square. Specifically, we consider the linear system of equations
+
+  a + b = (2 * l - 1) ^ 2
+  a + c = (2 * l) ^ 2
+  b + c = (2 * l + 1) ^ 2
+
+which can be solved to give
+
+  a = 2 * l * l - 4 * l
+  b = 2 * l * l + 1
+  c = 2 * l * l + 4 * l
+
+Therefore, it is enough to show that there exists a natural number l such that
+n ≤ 2 * l * l - 4 * l and 2 * l * l + 4 * l ≤ 2 * n for n ≥ 100.
+-/
+
+lemma exists_triplet_summing_to_squares (n : ℕ) (hn : n ≥ 100): (∃ (a b c : ℤ),
+(n : ℤ) ≤ a ∧ a < b ∧ b < c ∧ c ≤ 2 * n ∧  (∃ (k : ℤ), a + b = k * k) ∧
+(∃ (l : ℤ), a + c = l * l) ∧ (∃ (m : ℤ), b + c = m * m)) :=
+begin
+  --If n ≥ 107, we do not explicitly construct the triplet but use an existence
+  --argument from lemma above.
+  by_cases p : n ≥ 107,
+  have h := exists_numbers_in_interval n p,
+  cases h with l hl,
+  use (2 * l * l - 4 * l),
+  use (2 * l * l + 1),
+  use (2 * l * l + 4 * l),
+  refine ⟨_,⟨_,⟨_,⟨_,⟨_, ⟨_, _⟩⟩⟩⟩⟩⟩,
+  { rw le_sub_iff_add_le,
+    exact hl.2.1 },
+  { nlinarith },
+  { nlinarith },
+  { exact hl.2.2},
+  { use (2 * l - 1),
+    ring_exp_eq },
+  { use (2 * l),
+    ring_exp_eq },
+  { use(2 * l + 1),
+    ring_exp_eq },
+    --Otherwise, if 100 ≤ n < 107, then it suffices to consider explicit
+    --construction of a triplet {a, b, c}, which is constructed by setting l=9
+    --in the argument above this lemma.
+  { use 126,
+    use 163,
+    use 198,
+    repeat {split, linarith},
+    refine ⟨_,⟨_,_⟩⟩,
+    use 17,
+    linarith,
+    use 18,
+    linarith,
+    use 19,
+    linarith },
+end
+
+lemma mem_compl_iff_nmem_subset {A : finset ℕ} {a b c n : ℕ}
+(t : A ⊆ finset.range(2 * n + 1)\finset.range n) (h₁ : (n : ℤ) ≤ a)
+(h₂ : (a : ℤ) < b) (h₃ : (b : ℤ) < c) (h₄ : (c : ℤ) ≤ 2 * n) :
+(a ∉ A ↔ a ∈ (finset.range (2 * n + 1)\finset.range n)\A)
+∧ (b ∉ A ↔ b ∈ (finset.range (2 * n + 1)\finset.range n)\A) ∧
+(c ∉ A ↔ c ∈ (finset.range (2 * n + 1)\finset.range n)\A) :=
```

Comment:
formatting in this lemma needs to improve - spaces between the `\`(sdiffs) for example